### PR TITLE
Frontend: Basic redux

### PR DESCRIPTION
**Pull request by kobim**
Replacing the restaurants loading fetch with redux actions.

### CODE REVIEW

File: client/src/components/RestaurantsList/redux/actions.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+export const loadingFinished = list => ({
+  type: LOADING_FINISHED,
+  list,
+});
+
+export const loadingFailed = error => ({
+  type: LOADING_FAILED,
+  error,
+});
```
Comment (GeReV) [inline client/src/components/RestaurantsList/redux/actions.js]: I really recommend [`redux-actions`](https://redux-actions.js.org) for these.

File: client/src/components/RestaurantsList/redux/actions.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+export const fetchRestaurants = () => dispatch => {
+  dispatch(loadingStarted());
+  return fetch(`${API_URL}/restaurants`)
+    .then(response => response.json())
+    .then(list => dispatch(loadingFinished(list)))
+    .catch(err => dispatch(loadingFailed(err.message)));
+};
```
Comment (GeReV) [inline client/src/components/RestaurantsList/redux/actions.js]: I also recommend [`redux-api-middleware`](https://github.com/agraboso/redux-api-middleware) for this.

File: client/src/components/RestaurantsList/redux/actions.test.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+const middleware = [thunk];
+const mockStore = configureStore(middleware);
+
+describe('RestaurantsList actions', () => {
+  const store = mockStore({});
+  const fetchResponse = jest.fn();
+  let fetch;
+
+  beforeEach(() => {
+    fetch = global.fetch;
```
Comment (GeReV) [inline client/src/components/RestaurantsList/redux/actions.test.js]: There's a good mocking library for `fetch`.

File: client/src/components/RestaurantsList/redux/reducers.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  case LOADING_STARTED:
+    return { ...state, loading: true };
+  case LOADING_FAILED:
+    return { ...state, loading: false, error: action.error };
+  case LOADING_FINISHED:
+    return { ...state, loading: false, list: action.list };
+  default:
+    return state;
+  }
+};
```
Comment (GeReV) [inline client/src/components/RestaurantsList/redux/reducers.js]: `redux-actions` has a great function named `handleActions` for creating reducers.

File: client/src/components/RestaurantsList/redux/index.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { handleActions } from 'redux-actions';
+import { RSAA } from 'redux-api-middleware';
+
+import { API_URL } from '../../../config';
+
+export const LOADING_STARTED = 'weeat/restaurants/LOADING_STARTED';
```
Comment (NoyaArie) [inline client/src/components/RestaurantsList/redux/index.js]: FETCH_RESTAURANTS, FETCH_RESTAURANTS_SUCCESS, FETCH_RESTAURANTS_FAIL

File: client/src/components/RestaurantsList/redux/index.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+export const LOADING_FINISHED = 'weeat/restaurants/LOADING_FINISHED';
+export const LOADING_FAILURE = 'weeat/restaurants/LOADING_FAILURE';
+
+const INITIAL_STATE = {
```
Comment (NoyaArie) [inline client/src/components/RestaurantsList/redux/index.js]: initialState

File: client/src/components/RestaurantsList/index.jsx [hunk 7]
```diff
@@ -47,4 +32,18 @@ class RestaurantList extends Component {
...
-export default RestaurantList;
+RestaurantList.propTypes = {
+  list: PropTypes.array.isRequired,
+  loading: PropTypes.bool.isRequired,
+  error: PropTypes.string,
+
+  fetchRestaurants: PropTypes.func.isRequired,
+};
+
+const mapStateToProps = ({ restaurants }) => ({ ...restaurants });
```
Comment (NoyaArie) [inline client/src/components/RestaurantsList/index.jsx]: why not using selectors?

File: client/src/components/RestaurantsList/redux/index.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+const initialState = {
+  list: [],
+  loading: false,
+  error: null,
+};
+
+export const fetchRestaurants = () => ({
+  [RSAA]: {
+    types: [FETCH_RESTAURANTS, FETCH_RESTAURANTS_SUCCESS, FETCH_RESTAURANTS_FAIL],
+    endpoint: `${API_URL}/restaurants`,
```
Comment (NoyaArie) [inline client/src/components/RestaurantsList/redux/index.js]: out of scope - env file?

File: client/src/components/RestaurantsList/index.jsx (hunk 7 above)
Comment (kobim) [reply client/src/components/RestaurantsList/index.jsx]: saving for filters.

**Pull request merged**